compiler/rustc_mir_transform/src/dest_prop.rs

```diff
@@ -353,244 +366,300 @@ impl<'tcx> MutVisitor<'tcx> for Replacer<'tcx> {
-
-            // Since liveness is a backwards analysis, we need to walk the results backwards. To do
-            // that, we first collect in the `MaybeInitializedLocals` results in a forwards
-            // traversal.
-
-            live_and_init_locals.resize_with(data.statements.len() + 1, || {
-                BitSet::new_empty(body.local_decls.len())
-            });
-
-            // First, go forwards for `MaybeInitializedLocals` and apply intra-statement/terminator
-            // conflicts.
+        for (block, data) in traversal::preorder(self.body) {
+            // We need to observe the dataflow state *before* all possible locations in the CFG
+            // (statement or terminator) in each basic block, and then observe the state *after* the
+            // terminator. As long as `borrowed` does not have a "before" effect, we will observe
+            // all possible dataflow states.
+
             for (i, statement) in data.statements.iter().enumerate() {
-                this.record_statement_conflicts(statement);
+                self.apply_statement_conflicts(statement);
 
                 let loc = Location { block, statement_index: i };
-                init.seek_before_primary_effect(loc);
-
-                live_and_init_locals[i].clone_from(init.get());
+                // Before `statement` in the CFG is after the primary effect is applied, since this
+                // is a backwards analysis.
+                live.seek_after_primary_effect(loc);
+                self.apply_dataflow_conflicts(live.get());
             }
 
-            this.record_terminator_conflicts(data.terminator());
             let term_loc = Location { block, statement_index: data.statements.len() };
-            init.seek_before_primary_effect(term_loc);
-            live_and_init_locals[term_loc.statement_index].clone_from(init.get());
-
-            // Now, go backwards and union with the liveness results.
-            for statement_index in (0..=data.statements.len()).rev() {
-                let loc = Location { block, statement_index };
-                live.seek_after_primary_effect(loc);
+            live.seek_after_primary_effect(term_loc);
+            self.apply_dataflow_conflicts(live.get());
 
-                live_and_init_locals[statement_index].intersect(live.get());
+            self.apply_terminator_conflicts(data.terminator(), live.get());
 
-                trace!("record conflicts at {:?}", loc);
-
-                this.record_dataflow_conflicts(&mut live_and_init_locals[statement_index]);
-            }
-
-            init.seek_to_block_end(block);
             live.seek_to_block_end(block);
-            let mut conflicts = init.get().clone();
-            conflicts.intersect(live.get());
-            trace!("record conflicts at end of {:?}", block);
-
-            this.record_dataflow_conflicts(&mut conflicts);
+            self.apply_dataflow_conflicts(live.get());
```

Comment:
Preexisting, but shouldn't this visit the results in other direction?

The results are pre-calculated for each block entry point (start of a block for forward analysis, and end of a block for backward analysis). In addition to that the cursor stores the results for the current position.  Results for the current cursor position can be updated incrementally, but only when moving in the same direction as the analysis. Otherwise computation has to be restarted from the entry to the block.